cirq/google/engine/engine.py

```diff
@@ -152,47 +156,43 @@ class Engine:
     def __init__(self,
-                 api_key: Optional[str] = None,
-                 api: str = 'quantum',
+                 project_id: str,
                  version: str = 'v1alpha1',
-                 default_project_id: Optional[str] = None,
                  discovery_url: Optional[str] = None,
                  default_gcs_prefix: Optional[str] = None,
                  proto_version: ProtoVersion = ProtoVersion.V1,
                  **kwargs) -> None:
         """Engine service client.
 
+        Requires project_id.
+
         Args:
-            api_key: API key to use to retrieve discovery doc.
-            api: API name.
+            project_id: A project_id string of the Google Cloud Project to use.
+                API interactions will be attributed to this project and any
+                resources created will be owned by the project.
             version: API version.
-            default_project_id: A fallback project_id to use when one isn't
-                specified in the JobConfig given to 'run' methods.
-                See JobConfig for more information on project_id.
-            discovery_url: Discovery url for the API. If not supplied, uses
-                Google's default api.googleapis.com endpoint.
+            discovery_url: Discovery url for the API to select a non-default
+                backend for the Engine. If supplied, the `version` argument will
+                by ignored.
             default_gcs_prefix: A fallback gcs_prefix to use when one isn't
                 specified in the JobConfig given to 'run' methods.
                 See JobConfig for more information on gcs_prefix.
         """
-        self.api_key = api_key
-        self.api = api
-        self.default_project_id = default_project_id
-        self.version = version
-        self.discovery_url = discovery_url or ('https://{api}.googleapis.com/'
-                                               '$discovery/rest'
-                                               '?version={apiVersion}')
+        self.project_id = project_id
+        self.discovery_url = discovery_url or discovery.V2_DISCOVERY_URI
         self.default_gcs_prefix = default_gcs_prefix
         self.proto_version = proto_version
 
-        discovery_service_url = self.discovery_url
-        if self.api_key is not None:
-            discovery_service_url += '&key={}'.format(self.api_key)
-        self.service = discovery.build(
-            self.api,
-            self.version,
-            discoveryServiceUrl=discovery_service_url,
-            **kwargs)
+        request_builder = user_project_header_request_builder(project_id)
+        service_args = {
+            'requestBuilder': request_builder,
+        }
+        service_args.update(kwargs)
+
+        self.service = discovery.build('' if discovery_url else 'quantum',
+                                       '' if discovery_url else version,
+                                       discoveryServiceUrl=self.discovery_url,
```

Comment:
Just do `requestBuilder=request_builder, **kwargs`, instead of making a new dictionary and folding in the kwargs.

Reply:
I did it this way to allow the caller to override the requestBuilder. I think that the recommendation would break that ability since the requestBuilder argument would be specified twice. WDYT?

Reply:
Changed the format to add iff requestBuilder is not already defined on service_args.